Use "Set Password" page title for lost-password endpoint when setting a password

haszari opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

This PR has two changes:

- The page title of the `my-account/lost-password` endpoint can be adjusted based on the `action` URL param.
- The lost password reset form can now be used to set a password when the user is logged in.

The `Lost password` reset form can be used for various scenarios:

- User has forgotten or lost their password.
- Brand new account, user needs to set an initial password.

#### Allow changing endpoint/page title to `Set password` for new account flow
In the checkout block we're implementing a new create account (signup) flow where the user is sent a email notification and then clicks a link to set their new account password. In this scenario, it's not ideal to show the page title "Lost password"; it's a bit jarring and could be confusing. 

This PR extends the method (`get_endpoint_title()`) which generates page titles so an optional `action` parameter can be passed. For `lost-password` endpoint if the action is related to setting a password, different page title is returned - `Set password`. This action param is also passed to the relevant hook (`woocommerce_endpoint_{$endpoint}_title`) so this can be customized or overridden.

Wherever `get_endpoint_title()` is called, the `action` url param is passed through. Note that [`action=rp` or `action=resetpass` are conventionally used in WordPress core](https://github.com/WordPress/WordPress/blob/1079647d30f6cc5eab56c992e0a5f508426439b0/wp-login.php#L912), so I've added support for them, and also a new `newaccount` which will be passed from checkout block signup flow (in an open PR: https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/3236).

#### Allow `Set password` in logged-in browser session (for new account flow)
In the checkout signup flow, the user is logged in when the account is created, and then sent an email with the password reset link. It's likely the user will open the reset link in a browser where they are logged in (i.e. where they just completed the order). 

So as part of this change, the reset form is now shown even if the user is logged in. Previously it was not shown unless no user was logged in.

To protect against accidentally changing another user's password, there's an extra check added before redirecting to set-password form. If a user is logged in, and they don't match the user in the reset key, the password reset is aborted, with a notice. This is an edge case but important to prevent any chance of setting the password for the wrong user.

<img width="1421" alt="Screen Shot 2020-10-14 at 12 29 13 PM" src="https://user-images.githubusercontent.com/4167300/95928626-59015780-0e1e-11eb-9a6d-093368d9773c.png">

Also, as a side effect, if a user does a password reset, they can now do the last step in a browser where they happen to be already logged in. This might be a common case, when the user is logged in but has forgotten their password and needs to reset.

Closes #27754 .

### Screenshots

<img width="1306" alt="Screen Shot 2020-10-08 at 9 46 17 AM" src="https://user-images.githubusercontent.com/4167300/95385925-1e457e00-094b-11eb-98c8-29940cb33c68.png">

### How to test the changes
There are two aspects to test:

- Confirm the new signup flow from checkout block works as expected. **Note: this requires running a custom branch of WooCommerce Blocks, and running a dev build of blocks (checkout signup is disabled in production builds - feature gated).**
- Confirm that other uses of the `lost-password` endpoint work correctly and aren't negatively impacted.

#### Test new signup flow from checkout block
0. Ensure you can see emails from your site (e.g. set up MailHog).
1. Checkout blocks branch from https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/3236 and build, ensure is mapped into your test site as a plugin and is active. 
2. Add a checkout block to a page. Ensure you have a valid payment method set up.
3. Add something to cart and proceed to checkout block page.
3. Enter an email address and complete the form. 
3. Check the `Create an account` checkbox. (If you don't see it, ensure you are running a dev build of WooCommerce Blocks.)
2. Complete & submit checkout.
1. Find the `Your {store} account has been created!` email. Copy the `Click here to set your new password.` url.
    - Link should look something like this: `http://localhost:8333/my-account/lost-password/?action=newaccount&key=6lye4PPX11pbjWPJozSR&login=bob`
4. In an incognito window, or after logging out, navigate to set password url. 
6. Should see reset password form (2 password fields) with `Set password` title. 

Bonus points:

- Test using the reset password link in browsers with different users logged in:
   - Anon / no user.
   - Different user (from the new shopper account).
- Ensure the password reset only works if used with no logged in user or a browser where the same user is already logged in.

#### Test other uses of the `lost-password` endpoint
Test various password related flows - forgot/lost password. Whenever a my-account endpoint is rendered, ensure the page title is correct and makes sense (and the breadcrumb title matches). 

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

### Changelog entry

> Enhancement: Use more relevant `Set password` title for lost password reset form when applicable.

haszari committed:
support overriding/customising password reset page title via `action`:
- get_endpoint_title now takes an extra `action` param
  - this also is passed to the relevant hook (as an additional arg)
  - woocommerce_endpoint_{$endpoint}_title
- for `lost-password?action=rp`, use `Set password`
- pass action query param through when using get_endpoint_title

haszari committed:
use "Set password" title for lost-password endpoint for more actions:
- 'rp' and 'resetpass' are used in WordPress core
https://github.com/WordPress/WordPress/blob/1079647d30f6cc5eab56c992e0a5f508426439b0/wp-login.php#L913
- using 'newaccount' for checkout signup case

haszari committed:
improve PHPdocs for get_endpoint_title and related hook:
- add @since (original since based on git blame)
- document all params
- add @since for newly-added params
- add docs for woocommerce_endpoint_{$endpoint}_title

haszari commented:
I've optimistically added this to 4.7.0 milestone – hopefully it can land in that release and then my version check in blocks won't need to be adjusted.

haszari commented:
Note – this PR changes public interfaces (e.g. `get_endpoint_title()` and `woocommerce_endpoint_{$endpoint}_title`). Let me know what I need to do to document this and ensure it's communicated in the changelog, dev notes, etc.

vedanshujain reviewed (dismissed):
Works well and code looks good! I have tested existing lost password flows and modifying the `action` param to see what happens, but not the full flow from the blocks side, so a review from someone from the blocks team as well will be appreciated.

haszari committed:
show lost-password form when logged-in user:
- in the checkout signup use case, the user may be setting their
password in a logged-in browser session; they need to be able to set an
initial account password

haszari commented:
Thanks for the review @vedanshujain ! @Aljullu is reviewing the blocks PR so he can review this too.

Note that I've made one more change since you reviewed @vedanshujain - https://github.com/woocommerce/woocommerce/pull/27898/commits/2099ae8fc24080a0b235641ceb48216e52d5bb82. I've broadened this if condition so the `lost-password` reset form is shown even if the user is logged in. In the checkout create account case, the user is logged in as part of checkout, so when they click the set password link from email, they may be logged in. Without this change, the my-account page was showing, and the user wasn't able to set a password. 🙀 

Can you review this most recent change too and make sure there's no negative impact to my account, and set-password form ? Thanks again 😃 

Arguably this is an improvement. In `master` if you visit `/my-account/lost-password/` you get the dashboard, with a mismatched title; you have to log out to get the reset "enter email" form. With this change the user can trigger a password reset even if they are logged in.

<img width="1309" alt="Screen Shot 2020-10-09 at 2 51 47 PM" src="https://user-images.githubusercontent.com/4167300/95533185-f7687400-0a3e-11eb-9128-7f4d1a49ba95.png">

Aljullu reviewed (dismissed):
I tested the flow with WC Blocks and it worked like a charm. :ok_hand: I also took a quick look at the code without doing an in-depth review, and I couldn't spot anything. So approving from my side.

haszari commented:
Requesting another Proton review of this PR as I've changed the behaviour since this was reviewed. The `my-account/lost-password` endpoint reset form will now be shown **even if the user is logged in** (new behaviour). This is needed so that new signups can set a password (they are logged in by the checkout `Create account` logic).

I don't think this will cause any issues as the reset form is only shown if the reset key is valid. This also has a nice side effect that if a user triggers a lost-password flow when logged in, it will work fine. Previously the reset-password form would only be rendered if no user was logged in.

@vedanshujain can you re-review when you get a chance? Or volunteer someone else if you prefer. Thank you 😄

vedanshujain commented on includes/shortcodes/class-wc-shortcode-my-account.php:
```diff
@@ -39,7 +39,7 @@ public static function output( $atts ) {
 		}
 
-		if ( ! is_user_logged_in() ) {
+		if ( ! is_user_logged_in() || isset( $wp->query_vars['lost-password'] ) ) {
```
I think we should add another check, perhaps inline 54 or inside the lost password method, to make sure that if the user is logged in, then it's the same user who is trying to reset the password. This is to make it harder for any external plugin to accidentally switch users in the password reset action in case a user clicks on a password reset link while logged in as a different user. wdyt?

haszari replied:
That would be extra safe, I like this idea. Will take a look at that next :)

haszari replied:
Good catch! I tested this and it was possible to set another user's password. (Yikes!)

I tried two different fixes for this (see commit history if you're curious). 

The fix I went with aborts the password reset as early as possible, before redirecting to the set-password form 5e3713da75b68fa46874309b56f2947ab93134e0. I've also added a notice for the rare case that a user tries to use a password reset link in a browser session logged in as a different user.

Thanks for catching this @vedanshujain. Can you give it another review / test to be sure? Thanks again 🙇

haszari replied:
By the way, if you think it's necessary, we can add the check in `lost_password` as well, then the code is even more robust.

haszari replied:
I've added more info to the description about this change + a screenshot of the notice for clarity.

vedanshujain reviewed (commented):
Looks good other than a minor comment

haszari committed:
ensure lost-password form only resets password for correct user:
- if a user is logged in, check their ID matches the ID in the reset
  key / token

haszari committed:
check password-set user token before redirecting to set password form:
- this prevents the password reset process earlier (before the redirect)
- also now shows a notice informing the user that they need to log out
  of (other) account

haszari committed:
restore original comment (related changes are gone)

vedanshujain reviewed (approved):
New checks look good so approving. I also checked that we don't allow resetting password anyway unless you also have access to the reset password mail since we check whether the passsword reset cookie is also set or not.

haszari commented:
Thanks for helping get this shipshape @vedanshujain 🚢 🥳

haszari commented:
Added draft testing instructions: https://github.com/woocommerce/woocommerce/wiki/Release-Testing-Instructions-WooCommerce-4.7#improvements-to-my-account--lost-password-endpoint
